src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -0,0 +1,289 @@
+            if (mediaTypeList == null || mediaTypeList.Count == 0 || string.IsNullOrEmpty(contentType))
+            {
+                return false;
+            }
+
+            var mediaType = new MediaTypeHeaderValue(contentType);
+            foreach (var type in mediaTypeList)
+            {
+                if (mediaType.IsSubsetOf(type))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void FilterHeaders(List<KeyValuePair<string, object?>> keyValues, IHeaderDictionary headers, ISet<string> allowedHeaders)
+        {
+            foreach (var (key, value) in headers)
+            {
+                if (!allowedHeaders.Contains(key))
+                {
+                    // Key is not among the "only listed" headers.
+                    keyValues.Add(new KeyValuePair<string, object?>(key, "X"));
+                    continue;
+                }
+                keyValues.Add(new KeyValuePair<string, object?>(key, value.ToString()));
+            }
+        }
+
+        private async Task<string> ReadRequestBody(HttpRequest request, CancellationToken token)
+        {
+            if (_options.BodyEncoding == null)
+            {
+                return "X";
+            }
+
+            using var joinedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token);
+            joinedTokenSource.CancelAfter(_options.RequestBodyTimeout);
+            var limit = _options.RequestBodyLogLimit;
+
+            // Use a pipe for smaller payload sizes
+            if (limit <= PipeThreshold)
+            {
+                try
+                {
+                    while (true)
+                    {
+                        // TODO if someone uses the body after this, it will not have the rest of the data.
+                        var result = await request.BodyReader.ReadAsync(joinedTokenSource.Token);
+                        if (!result.IsCompleted && result.Buffer.Length <= limit)
+                        {
+                            // Need more data.
+                            request.BodyReader.AdvanceTo(result.Buffer.Start, result.Buffer.End);
+                            continue;
+                        }
+
+                        var res = _options.BodyEncoding.GetString(result.Buffer.Slice(0, result.Buffer.Length > limit ? limit : result.Buffer.Length));
+
+                        request.BodyReader.AdvanceTo(result.Buffer.Start, result.Buffer.End);
+                        return res;
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    // Token source hides triggering token (https://github.com/dotnet/runtime/issues/22172)
+                    if (!token.IsCancellationRequested && joinedTokenSource.Token.IsCancellationRequested)
+                    {
+                        return "X";
```

Comment:
Currently, any failure causes the log to write "X". Technically, I could write a different string to the logs here, like [Canceled].